### PR TITLE
179438 - Handle Successful Empty Response for UPNs and Holdings

### PR DESCRIPTION
**Pull request by shailen14599**
This PBI consists changes of [PBI 179438](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/179438)

Added code changes 
Added unit tests
Added functional tests

### CODE REVIEW

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.API/Controllers/PermitController.cs [hunk 1]
```diff
@@ -29,12 +30,12 @@ public PermitController(IHttpContextAccessor httpContextAccessor,
         public virtual async Task<IActionResult> GeneratePermits(int licenceId)
         {
             _logger.LogInformation(EventIds.GeneratePermitStarted.ToEventId(), "Generate Permit API call started.");
-            
-            await _permitService.CreatePermitAsync(licenceId, GetRequestCancellationToken(), GetCorrelationId());
+
+            var httpStatusCode = await _permitService.CreatePermitAsync(licenceId, GetRequestCancellationToken(), GetCorrelationId());
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.API/Controllers/PermitController.cs:34]:
```suggestion
            var responseStatusCode = await _permitService.CreatePermitAsync(licenceId, GetRequestCancellationToken(), GetCorrelationId());
```

File: src/UKHO.S100PermitService.API/Controllers/PermitController.cs
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.API/Controllers/PermitController.cs:1]: minor inconsistent formatting on constructor, please fix it.

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 2]
```diff
@@ -31,15 +33,28 @@ public PermitService(IPermitReaderWriter permitReaderWriter,
...
+            string correlationId)
         {
             _logger.LogInformation(EventIds.CreatePermitStart.ToEventId(), "CreatePermit started");
 
+            var userPermitServiceResponse = await _userPermitService.GetUserPermitAsync(licenceId, cancellationToken, correlationId);
+
+            if(IsUserPermitServiceResponseNull(userPermitServiceResponse))
+            {
+                _logger.LogWarning(EventIds.UserPermitServiceGetUserPermitsRequestCompletedWithNoContent.ToEventId(), "Request to UserPermitService responded with empty response");
+                return HttpStatusCode.NoContent;
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:46]:
```suggestion
                _logger.LogWarning(EventIds.UserPermitServiceGetUserPermitsRequestCompletedWithNoContent.ToEventId(), "Request to UserPermitService responded with empty response");
                
                return HttpStatusCode.NoContent;
```

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 3]
```diff
@@ -31,15 +33,28 @@ public PermitService(IPermitReaderWriter permitReaderWriter,
...
+            }
+
             var holdingsServiceResponse = await _holdingsService.GetHoldingsAsync(licenceId, cancellationToken, correlationId);
 
-            var productsList = GetProductsList();
+            if(IsListEmptyOrNull(holdingsServiceResponse))
+            {
+                _logger.LogWarning(EventIds.HoldingsServiceGetHoldingsRequestCompletedWithNoContent.ToEventId(), "Request to HoldingsService responded with empty response");
+                return HttpStatusCode.NoContent;
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:54]:
```suggestion
                _logger.LogWarning(EventIds.HoldingsServiceGetHoldingsRequestCompletedWithNoContent.ToEventId(), "Request to HoldingsService responded with empty response");
                
                return HttpStatusCode.NoContent;
```

File: src/UKHO.S100PermitService.StubService/StubData/UserPermits/response-200-licenceid-6.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "licenceId": "1",
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.StubService/StubData/UserPermits/response-200-licenceid-6.json:2]:
```suggestion
  "licenceId": 1,
```

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs [hunk 5]
```diff
@@ -159,5 +172,130 @@ public async Task WhenEmptyPermitXml_ThenFileIsNotCreated()
...
+
+                case NoContent:
+                    return
+                    [
+                    ];
+
+                default:
+                    return null;
+            }
+
```
[the remark this answers is not in the document]
Comment (shailen14599) [reply tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs]: Resolved

File: src/UKHO.S100PermitService.API/Controllers/PermitController.cs (hunk 1 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.API/Controllers/PermitController.cs]: Resolved

File: src/UKHO.S100PermitService.API/Controllers/PermitController.cs
Comment (shailen14599) [reply src/UKHO.S100PermitService.API/Controllers/PermitController.cs:1]: Resolved

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 2 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 3 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: Resolved

File: src/UKHO.S100PermitService.StubService/StubData/UserPermits/response-200-licenceid-6.json (hunk 4 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.StubService/StubData/UserPermits/response-200-licenceid-6.json]: Resolved

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

**Review by harshal11869 — APPROVED** [review]
Looks Good.

File: src/UKHO.S100PermitService.Common/Extensions/ListExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace UKHO.S100PermitService.Common.Extensions
+{
+    [ExcludeFromCodeCoverage]
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Extensions/ListExtensions.cs:5]: Should really write a quick set of unit tests for this

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 7]
```diff
@@ -108,11 +128,12 @@ private static List<Products> GetProductsList()
...
         }
 
-        private static List<ProductKeyServiceRequest> ProductKeyServiceRequest(List<HoldingsServiceResponse> holdingsServiceResponse) =>
-             holdingsServiceResponse.SelectMany(x => x.Cells.Select(y => new ProductKeyServiceRequest
-             {
-                 ProductName = y.CellCode,
-                 Edition = y.LatestEditionNumber
-             })).ToList();
+        private static List<ProductKeyServiceRequest> ProductKeyServiceRequest(
+            List<HoldingsServiceResponse> holdingsServiceResponse) =>
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:132]: Generic List(T) argument should be IEnumerable(T)

**Review by jrippington — CHANGES_REQUESTED** [review]
Couple of minor points, thanks!

File: src/UKHO.S100PermitService.Common/Extensions/ListExtensions.cs (hunk 6 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Extensions/ListExtensions.cs]: Added unit tests for ListExtensions class

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 7 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: Resolved

**Review by Arit14975 — APPROVED** [review]
LGTM from testing side.

File: src/UKHO.S100PermitService.Common/Validations/UserPermitServiceResponseValidator.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using UKHO.S100PermitService.Common.Models.UserPermitService;
+
+namespace UKHO.S100PermitService.Common.Validations
+{
+    [ExcludeFromCodeCoverage]
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Validations/UserPermitServiceResponseValidator.cs:6]: This needs a unit test

**Review by jrippington — CHANGES_REQUESTED** [review]
Few tests needed, thanks!

File: src/UKHO.S100PermitService.Common/Validations/UserPermitServiceResponseValidator.cs (hunk 8 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Validations/UserPermitServiceResponseValidator.cs]: Resolved. Added unit tests

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Pull request merged**